ios notifs: Consume "background queue" so notifs navigate from backgr…

chrisbobbe opened the pull request:
…ound.

This is a short-term fix that relies on a deprecated API. We should
upgrade to a newer version of react-native-notifications or use
@react-native-community/push-notification-ios instead.

For now-outdated docs on this, see
https://github.com/wix/react-native-notifications/commit/fe30ae5e03990eb46dc7c04a9ebe86ae0338c7d6#diff-d455ef90f34ba49d565c96e9f2cc8b98R53-R58.
The API was removed in release 2.0.6-shapshot-8 with commit
https://github.com/wix/react-native-notifications/commit/800669120bb7d5f5220ad821a323c3ab7f94342f,
presumably in response to
https://github.com/wix/react-native-notifications/issues/339.

Process an opened notification from the "background queue" so our
handleNotificationOpen will run, and the app will navigate, for a
notification opened while the app was in the background.

Fixes: #3647

chrisbobbe commented:
If, soon, we can use a later version of react-native-notifications, or switch to @react-native-community/push-notification-ios, then this PR should be closed in favor of that solution.

chrisbobbe commented on src/boot/AppEventHandlers.js:
```diff
@@ -95,6 +96,17 @@ class AppEventHandlers extends PureComponent<Props> {
+    if (state === 'active' && Platform.OS === 'ios') {
+      // TODO: This API is deprecated in react-native-notifications release
+      // 2.0.6-snapshot.8; see #3647.
+      try {
```
See motivation for try/catch at https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/iOS.20push.20notifications/near/825135.

ray-kraesig replied:
I dug into this a bit – but, unfortunately, not enough to have a solid answer to that question.

I have learned that it probably *shouldn't* throw. The [definition of `consumeBackgroundQueue`](https://github.com/wix/react-native-notifications/blob/v1.4.0/RNNotifications/RNNotifications.m#L549) uses [`RCT_EXPORT_METHOD`](https://github.com/facebook/react-native/blob/v0.59.10/React/Base/RCTBridgeModule.h#L134-L175), which creates [an asynchronous method](https://github.com/facebook/react-native/blob/v0.59.10/React/Base/RCTBridgeModule.h#L208-L210) which [doesn't return any sort of value](https://github.com/facebook/react-native/blob/v0.59.10/React/Base/RCTModuleMethod.mm#L557-L562). (Even via callbacks, since it has no parameters.)

However, the invocation of a native method is eventually done [here](https://github.com/facebook/react-native/blob/v0.59.10/React/CxxModule/RCTNativeModule.mm#L91-L120), and the comment [here](https://github.com/facebook/react-native/blob/v0.59.10/React/CxxModule/RCTNativeModule.mm#L108) implies that some `NSException`s are propagated back to JS. I haven't found where that would happen.

ray-kraesig commented on src/boot/AppEventHandlers.js:
```diff
@@ -95,6 +96,17 @@ class AppEventHandlers extends PureComponent<Props> {
+        NotificationsIOS.consumeBackgroundQueue();
+      } catch (e) {
+        // Do nothing
+      }
```
At the very least we need to feed these errors to Sentry. In particular, if we upgrade `react-native-notifications` to a later version without fixing this call, it *will* start throwing, and we may never notice! (Flow has no functional typing information for this module, so won't complain.)

ray-kraesig commented:
> If, soon, we can use a later version of react-native-notifications, or switch to @react-native-community/push-notification-ios, then this PR should be closed in favor of that solution.

Not soon enough, I'm sorry to say. :frowning_face: Fortunately, this fix is pretty small and isolated, and shouldn't cause us any problems when upgrading or switching.

gnprice commented on src/boot/AppEventHandlers.js:
```diff
@@ -89,6 +91,22 @@ class AppEventHandlers extends PureComponent<Props> {
+            'Call to NotificationsIOS.consumeBackgroundQueue failed; pressed notification failed to navigate',
+        });
+      }
+    }
```
Let's move all this into its own function, in our notifications code in `src/notifications/index.js` -- I think that's a better home for this complexity to live, next to the other code that has to deal with the intricacies of these libraries. Perhaps right after `handleInitialNotification`, which has a similar job.

Probably the `Platform.OS === 'ios'` condition belongs inside that function, too. Then this can just look like
```js
  if (state === 'active') {
    notificationOnAppActive();
  }
```
or some other name that indicates this is the function for doing notification-related things that we do when the app moves into the foreground / into state `active`. (And the details of what those things are is the job of the function's implementation, as part of the notification code.)

chrisbobbe replied:
OK, sounds good!

chrisbobbe commented:
Thanks for the review! 🙂

> Have you tested this, and it seems to fix the issue?

Yes, I did. I'm setting up the dev server again right now, to test that these latest changes I've made haven't introduced bugs.

chrisbobbe commented:
OK, this should be ready to go. I changed the comments slightly; I put them inside the `try` block and added a bit about why the code is needed in the first place.

gnprice commented:
Thanks for the revisions! Merged.
